Firestore/core/src/firebase/firestore/remote/buffered_writer.h

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_REMOTE_BUFFERED_WRITER_H
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_REMOTE_BUFFERED_WRITER_H
+
+#include <functional>
+#include <vector>
+
+#include <grpcpp/support/byte_buffer.h>
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+// GRPC doesn't allow issuing a write operation before the previous write has
+// finished. This class helps keep track of pending writes and whether there
+// currently is a write operation in progress; it doesn't communicate with GRPC
+// directly and expects the following to function correctly:
+// - the constructor should be given a function that actually issues the GRPC
+//   write operation;
+// - each time a write operation finishes, `OnSuccessfulWrite` must be called.
+// The invariant is that `BufferedWriter` will never invoke the writing function
+// if `OnSuccessfulWrite` hasn't been called since the previous invocation.
+//
+// The main methods are `Enqueue` and `OnSuccessfulWrite`. `Enqueue` will invoke
+// the writing function immediately if there is no operation in progress;
+// otherwise, the given `bytes` will be stored in the buffer.
+// `OnSuccessfulWrite` will invoke the writing function with the next write in
+// buffer, as long as the buffer is not empty (in FIFO order).
+
+// `BufferedWriter` can be `Start`ed and `Stop`ped, which is expected to reflect
```

Comment:
I don't entirely understand why starting and stopping a BufferedWriter is actually useful. In our usage it seems as if attempting to add a message to a stopped stream is an error. Is this logic something we'd ever actually use?

Note that in `FSTBufferedWriter` the state is required to implement the `GRXWriter` protocol which gets callbacks from gRPC to actually implement the one-at-a-time flow control. Since in this implementation we're counting on the caller to call `OnSuccessfulWrite` it seems as if we're already doing flow control for ourselves.